chain/client/src/client_actor.rs

```diff
@@ -853,22 +855,32 @@ impl ClientActor {
         &mut self,
         block: Block,
         provenance: Provenance,
+        peer_id: &PeerId,
     ) -> Result<(), near_chain::Error> {
         // If we produced the block, send it out before we apply the block.
         // If we didn't produce the block and didn't request it, do basic validation
         // before sending it out.
         if provenance == Provenance::PRODUCED {
             self.network_adapter.do_send(NetworkRequests::Block { block: block.clone() });
-        } else if provenance == Provenance::NONE {
-            // Don't care about challenge here since it will be handled when we actually process
-            // the block.
-            if self.client.chain.process_block_header(&block.header(), |_| {}).is_ok() {
-                let head = self.client.chain.head()?;
-                // do not broadcast blocks that are too far back.
-                if head.height < block.header().height()
-                    || &head.epoch_id == block.header().epoch_id()
-                {
-                    self.client.rebroadcast_block(block.clone());
+        } else {
+            match self.client.chain.validate_block(&block) {
+                Ok(_) => {
+                    let head = self.client.chain.head()?;
+                    // do not broadcast blocks that are too far back.
+                    if head.height < block.header().height()
+                        || &head.epoch_id == block.header().epoch_id()
+                    {
+                        self.client.rebroadcast_block(block.clone());
```

Comment:
Won't you end up rebroadcasting blocks during sync?

Reply:
I am a n00b